pcsx2/gui/FrameForGS.cpp

```diff
@@ -243,6 +257,66 @@ void GSPanel::DoShowMouse()
 	m_HideMouseTimer.Start( 1750, true );
 }
 
+std::optional<WindowInfo> GSPanel::GetWindowInfo()
+{
+	WindowInfo ret;
+
+	const wxSize gs_vp_size(GetClientSize());
+	ret.surface_scale = static_cast<float>(GetContentScaleFactor());
+	ret.surface_width = static_cast<u32>(gs_vp_size.GetWidth());
+	ret.surface_height = static_cast<u32>(gs_vp_size.GetHeight());
```

Comment:
If we're tracking this info here, do you plan to call this every frame?  Otherwise how do you handle window resizes?  If you are, I don't know about other platforms, but on macOS these methods fall under the "must be called from the main thread" category, which makes calling it from the GS thread kind of difficult.

Reply:
At the moment, it's tracked by the GS thread, same as it is in master. This is just the initial dimensions, which are (currently) unused. Trying to avoid too many changes at once :)

For Wayland, it's impossible to query the size, so the size events need to be passed to the GS thread anyway. My plan was to add another command type to MTGS which passes the dimensions through, and resizes the target on the GL thread (which is safe).

For Macs, yeah, it's a pain. I haven't messed with it in a while, but for DuckStation I just use GCD to kick the GL context update back to the main thread. It's kinda janky to have GL stuff in the frontend/window side of things IMO.

Reply:
Also, for Macs, I can't see how it would even work at the moment given there's no window/GL context wrapper. Rather than implementing GSWnd and then tossing it away when I drop GSWnd completely, I was just going to leave that for later.

Reply:
> Also, for Macs, I can't see how it would even work at the moment given there's no window/GL context wrapper. 

I'm currently maintaining it on a fork as it requires some extra work to work around Apple's ancient OpenGL implementation, but you can see it here: https://github.com/tellowkrinkle/pcsx2/blob/MacOS/pcsx2/GS/Window/GSWndCGL.mm

Reply:
Seems pretty similar to what I did: https://github.com/stenzek/duckstation/blob/master/src/common/gl/context_agl.mm

Indeed, there's a potential race between the backbuffer actually updating (on the main thread) and the rendering thread using the new size (passed in through the resize event). But that's going to happen regardless, even if you did the resize on the main thread, unless you paused rendering while the update happens, which I feel like has the potential to deadlock. As far as I'm aware, the worst that would happen is you render to a smaller/invalid viewport for one frame.

I guess another option could be to block the rendering thread until the main thread processes the update, but I haven't tried that.

Reply:
> But that's going to happen regardless, even if you did the resize on the main thread, unless you paused rendering while the update happens, which I feel like has the potential to deadlock. As far as I'm aware, the worst that would happen is you render to a smaller/invalid viewport for one frame.

Apple doesn't talk about the thread safety of trying to access NSView parameters from anywhere other than the main thread, other than to say "just don't do it".  It's *probably* okay, but it makes their automated checker scream and throw warnings everywhere, which kind of bothers me so I'd rather avoid it if possible

And yeah I think having the chance of being a frame behind in your resize is kind of unavoidable unless you're doing your rendering from the main thread (which we're not)

Reply:
> Apple doesn't talk about the thread safety of trying to access NSView parameters from anywhere other than the main thread, other than to say "just don't do it". It's probably okay with these read-only things, but it makes their automated checker scream and throw warnings everywhere, which kind of bothers me so I'd rather avoid it if possible

With the way I'm envisioning things, there won't be any `NSView` calls happening on the GS thread, it'll all be on the UI thread, and then handed off to the GS thread. Which might result in redundant resizes if the user is dragging the window out, but it should be safe.